### PR TITLE
Default XSD/doc values are now extracted from source.

### PR DESCRIPTION
**Pull request by ryanemerson**
This consists of a java program `DefaultAttributeExtractor.java`, which outputs the default values for configuration resources found on the classpath to either a propteries file or asciidoc attributes file. The process of extracting default key/values combos are provided in the specified implementation of the DefaultsResolver interface. 

Values extracted to a defaults.proprties file, on relevant modules, are then used to apply a maven-resource filter to *.xsd files where ${...} placeholders are replaced with the relevant default values.  

For asciidoctor integration, the default values are exported to a .asciidoctor file where they are defined as attributes. To use these values you simply include the defaults.asciidoctor file and call the attribute where required, e.g. for the simple cache default can be called via `{Configuration.simpleCache}`.

@tristantarrant I have updated all of the relevant *.xsd files (AFAIK) to use default placeholders, however the asciidoc files remain the same as many of the defaults are referenced in sentences etc and need more attention.

### CODE REVIEW

**Comment by galderz** [comment]
Needs rebasing

File: persistence/soft-index/pom.xml [hunk 1]
```diff
@@ -81,6 +81,58 @@
...
+            </configuration>
+            <executions>
+               <execution>
+                  <id>defaults</id>
+                  <phase>process-classes</phase>
+                  <goals>
+                     <goal>java</goal>
+                  </goals>
+                  <configuration>
+                     <mainClass>org.infinispan.commons.configuration.defaults.DefaultAttributeExtractor</mainClass>
```
Comment (slaskawi) [inline persistence/soft-index/pom.xml]: Maybe we should create a Maven plugin for it? This parts is being copied all over the place.

File: commons/src/main/java/org/infinispan/commons/configuration/defaults/InfinispanDefaultsResolver.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.configuration.attributes.Attribute;
+import org.infinispan.commons.configuration.attributes.AttributeDefinition;
+import org.infinispan.commons.configuration.attributes.AttributeSet;
+
+/**
+ * @author Ryan Emerson
```
Comment (slaskawi) [inline commons/src/main/java/org/infinispan/commons/configuration/defaults/InfinispanDefaultsResolver.java]: Could you please remove those empty javadocs? Or even better - put something meaningful here?

File: persistence/soft-index/pom.xml (hunk 1 above)
Comment (ryanemerson) [inline persistence/soft-index/pom.xml]: @slaskawi Yeah I did think that a Maven plugin would maybe better once I started pasting the same stuff into the pom. I've never developed a plugin before, so I'll look into what that entails etc

**Comment by ryanemerson** [comment]
@slaskawi As suggested I have refactored the PR as a maven plugin, which has reduced the verbosity of the pom's somewhat, and I have improved the javadocs.

File: plugins/infinispan-defaults-maven-plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-plugin-plugin</artifactId>
+                <version>3.5</version>
+                <configuration>
+                    <skipErrorNoDescriptorsFound>true</skipErrorNoDescriptorsFound>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.infinispan</groupId>
```
Comment (slaskawi) [inline plugins/infinispan-defaults-maven-plugin/pom.xml]: This is probably a mistake - the plugin uses itself?

Comment (ryanemerson) [inline plugins/infinispan-defaults-maven-plugin/pom.xml]: @slaskawi Good spot :) I've pushed the changes

**Comment by galderz** [comment]
Anything else to do here?

**Comment by ryanemerson** [comment]
@galderz Not as far as I'm concerned

**Comment by slaskawi** [comment]
Hey @ryanemerson Looks good to me. Could you please rebase it. Once you and TeamCity are done on this PR, I'll integrate it.

**Comment by ryanemerson** [comment]
@slaskawi rebased, thanks!

**Comment by slaskawi** [comment]
Thanks @ryanemerson !

**Pull request merged**